### PR TITLE
feat: adds request limiter

### PR DESCRIPTION
**Pull request by VinceGhi**
I personally had to use a limiter quite often so i thought it might be usefull here as well to.. will also reduce the amount of "useless" requests towards torn. Is actually more usefull in combination with the MultiKeyProvider from the other PR.

Also added a option to make async api-calls which also is usefull quite often and kinda fits the limit.

### CODE REVIEW

**Comment by DeKleineKobini** [comment]
If I understand correctly, you wouldn't be able to use this with bursts of requests? 
For example if I have a cron job every minute to make 10 requests within a second, but only once that minute, this would fail all but the first request?

**Comment by VinceGhi** [comment]
It currently would sleep until the next api call would be allowed... should honestly be configurable if the user wants it to sleep or fail... 🤔

**Comment by VinceGhi** [comment]
The limiter had issues which i had to fix.

It basically just looks up how many requests have been made within the timeframe of the limiter and lets the thread sleep or throw an exception if the limiter has reached the exception.

So if you CronJob has a client with a limit of 1 request every minute it would queue them all and make the first request, realise that the limit has been reached at the next request and lets it sleep for 1min. The third request would be executed 2min after the first one, and so on.

Or you enable exceptions and switch key etc...

File: src/main/java/eu/tornplayground/tornapi/limiter/RequestLimiter.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+import java.util.Deque;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Limits the number of requests per given time frame per API key.
+ */
+public class RequestLimiter {
```
Comment (DeKleineKobini) [inline src/main/java/eu/tornplayground/tornapi/limiter/RequestLimiter.java:12]: I feel like this really should be an interface and this logic be moved to it's own class. This implementation spaces requests out evenly, but a different implementation could allow some more bursts of requests.

File: src/main/java/eu/tornplayground/tornapi/limiter/DefaultRequestLimiter.java [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package eu.tornplayground.tornapi.limiter;
+
+/**
+ * Default implementation of {@link RequestLimiter} with a limit of 100 requests per minute per API key.
+ * That's the default limit for the Torn API.
+ */
+public class DefaultRequestLimiter extends RequestLimiter {
```
Comment (DeKleineKobini) [inline src/main/java/eu/tornplayground/tornapi/limiter/DefaultRequestLimiter.java:7]: Not really a fan of implying this as the default, as there is no need to space out requests evenly by default.

File: src/main/java/eu/tornplayground/tornapi/limiter/RequestLimiter.java (hunk 1 above)
Comment (VinceGhi) [reply src/main/java/eu/tornplayground/tornapi/limiter/RequestLimiter.java:12]: You can still do bursts of requests tho.. if you have 100 requests per minute you can do 100 requests within the first second.. that doesn't matter.. it literally just checks if there has been X request within the given timeframe. But maybe i don't understand what you mean

Comment (DeKleineKobini) [reply src/main/java/eu/tornplayground/tornapi/limiter/RequestLimiter.java:12]: Oh yeah, now I understand. Was too fixated on the constructor for some reason.

**Review by DeKleineKobini — APPROVED** [review]
Looks good, just need the merge conflict fixed before merging.

**Pull request merged**